### DISCUSSION TITLE
How Laravelly does a codebase need to be for Larastan?

### DISCUSSION THREAD

**Discussion by adnweedon** [larastan/larastan#1613] [Q&A]
I've been doing a really fun project recently, integrating some Laravel features into an existing codebase, using the illuminate packages that are available. We've got an application container running, which is injecting dependencies into most of our endpoints now, which is great.

We use PHPStan to analyse our code, but I can see that getting less useful once we're using DI more. My question is, how much like Laravel does a system need to be for Larastan to work at all, and be useful? Is it enough to have a Illuminate\Contracts\Foundation\Application with some bootstrap code, or do other things also need to match Laravel for it to work? I realise it's a bit of a vague question in some ways, but I couldn't find any discussion about using Larastan on Laravel-esque systems!

**Comment by szepeviktor** [comment]
This is a very nice discussion.

**Comment by szepeviktor** [comment]
Larastan is a collection of PHPStan extensions.
https://github.com/nunomaduro/larastan/blob/master/extension.neon
You may pick and choose what fits your project.

You may want to take a look at our app resolver: https://github.com/nunomaduro/larastan/blob/master/src/ApplicationResolver.php#L59

**Reply by adnweedon** [reply]
Thanks for the links - you'll have to forgive me for my lack of knowledge, but I looked through those files, but nothing immediately stood out to me! I'm going to share my progress so far, in case it helps others, and to hopefully hone in on sensible questions. 

### Getting it running

So my first obstacle is to get Larastan running at all (having added the neon config); looking through the bootstrap file, I see that the absolute minimum is:

- having an application container that implements Illuminate\Contracts\Foundation\Application (or the Lumen version)
- a bootstrap file to fire it up located at bootstrap/app.php
- a console Kernel implementing Illuminate\Contracts\Console\Kernel, that the container can build

I already had the first one, so I shuffled the bootstrapping code I already had into a bootstrap/app.php file. I didn't have a console kernel, but from searching the Larastan codebase, I can't find it being used for anything other than bootstrapping, so I threw together the absolute basics, only properly implemented the bootstrap() method, and registered it as a singleton for the Interface. Voila - PHPStan runs again!

At this point, I am fairly confident that I have PHPStan running with the Larastan extensions loaded in; it's certainly doing something to boot the container, and is successfully doing an analysis.

### Next steps

So what I'd really like to do, as a first test, is have it detect when I'm trying to resolve something from the container that it can't build, and would give me a binding resolution exception (I presume this is something it can do - I implied it from the description on the front page, but couldn't see an obvious extension for it in the extension.neon file?). So, I added one to a method that definitely can't be resolved, and ran it - no luck though, I'm still not seeing any errors.

We're not yet using Laravel-style routing; it will probably come in the future, but for now we're just getting used to having a Service Container, and are using its call() method to inject our existing endpoints through our existing router (we've made a Kernel class that is handling bootstrapping, Laravel style, but it's not handling the request yet!). Is that the issue here, perhaps - is there some way to help Larastan identify methods that are being called by the container?
